cli/pull.ml

```diff
@@ -65,9 +70,25 @@ let run (`Yes yes) (`Root root) (`Lockfile explicit_lockfile)
         Common.filter_duniverse ~to_consider:duniverse_repos duniverse
       in
       let* () = check_dune_lang_version ~yes ~root in
-      OpamGlobalState.with_ `Lock_none (fun global_state ->
-          Pull.duniverse ~global_state ~root ~full
-            ~trim_clone:(not keep_git_dir) duniverse)
+      OpamGlobalState.with_ `Lock_none @@ fun global_state ->
+      let* locked_ocaml_version =
+        Lockfile.ocaml_version lockfile
+        |> Result.of_option ~error:(`Msg "OCaml compiler not in lockfile")
+      in
+      OpamSwitchState.with_ `Lock_none global_state @@ fun switch_state ->
+      let switch_ocaml_version =
+        OpamSwitchState.get_package switch_state Config.compiler_package_name
+        |> OpamPackage.version
+      in
+      (match version_is_at_least locked_ocaml_version switch_ocaml_version with
+      | true -> ()
+      | false ->
+          Logs.warn (fun l ->
+              l "OCaml versions do not match: %a in lockfile, but %a in switch"
```

Comment:
It might be useful to suggest how to fix this along with the diagnostic, how about something along those lines:
```
You might want to upgrade the compiler in your switch:
opam install ocaml.%a --update-invariant
```

Reply:
I didn't know this was possible on a switch with a locked invariant — that's a great idea, I added it!